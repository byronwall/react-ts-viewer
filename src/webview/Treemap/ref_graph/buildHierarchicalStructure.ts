import type { ScopeNode } from "../../../types";
import { findCommonAncestor } from "./graph_nodes";
import { getPathToNode } from "./graph_nodes";

// Helper function to build hierarchical structure preserving parent-child relationships

export function buildHierarchicalStructure(
  nodes: ScopeNode[],
  rootNode: ScopeNode
): ScopeNode {
  const nodeIds = nodes.map((n) => n.id);
  const commonAncestor = findCommonAncestor(rootNode, nodeIds);

  // Build the path from root to the common ancestor so we can include higher-level context
  const ancestorPath = getPathToNode(rootNode, commonAncestor.id);

  // Helper to attach child into shallow copy
  const cloneNode = (n: ScopeNode): ScopeNode => ({
    ...n,
    children: n.children ? [] : [],
  });

  // Build a tree that includes all necessary nodes and their hierarchical relationships
  function buildSubtree(node: ScopeNode): ScopeNode | null {
    // Check if this node should be included
    const shouldInclude = nodeIds.includes(node.id);

    // Process children recursively
    const includedChildren: ScopeNode[] = [];
    if (node.children && node.children.length > 0) {
      for (const child of node.children) {
        const builtChild = buildSubtree(child);
        if (builtChild) {
          includedChildren.push(builtChild);
        }
      }
    }

    // Include this node if either:
    // 1. It's explicitly in our target nodes, OR
    // 2. It has children that need to be included (intermediate node)
    if (shouldInclude || includedChildren.length > 0) {
      const result: ScopeNode = {
        ...node,
        children: includedChildren, // Always use the array, even if empty
      };
      return result;
    }

    return null;
  }

  const hierarchicalTree = buildSubtree(commonAncestor);

  // Now graft this subtree under the ancestor chain to reach the root
  let current: ScopeNode = hierarchicalTree || commonAncestor;
  for (let i = ancestorPath.length - 2; i >= 0; i--) {
    const anc = ancestorPath[i];
    if (!anc) continue;
    const parent = cloneNode(anc);
    parent.children = [current];
    current = parent;
  }

  return current;
}
