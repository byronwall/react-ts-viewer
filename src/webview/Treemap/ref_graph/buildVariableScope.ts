import * as ts from "typescript";
import { VariableScope } from "./graph_nodes";
import { getLineAndCharacter } from "./ts_ast";
import { extractDestructuredNames } from "./ts_ast";

// Build variable scope from AST

export function buildVariableScope(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  parent?: VariableScope
): VariableScope {
  const scope: VariableScope = {
    declarations: new Map(),
    parent,
    level: parent ? parent.level + 1 : 0,
  };

  const walk = (n: ts.Node): void => {
    // Variable declarations (simple identifiers)
    if (
      ts.isVariableDeclaration(n) ||
      ts.isParameter(n) ||
      ts.isFunctionDeclaration(n)
    ) {
      if (n.name && ts.isIdentifier(n.name)) {
        const name = n.name.text;
        const position = getLineAndCharacter(sourceFile, n.name.pos);
        scope.declarations.set(name, { node: n, name, line: position.line });
      }
    }

    // Destructuring patterns â€“ capture each individual name
    if (ts.isVariableDeclaration(n) && n.name) {
      extractDestructuredNames(n.name).forEach((name) => {
        const position = getLineAndCharacter(sourceFile, n.name!.pos);
        scope.declarations.set(name, { node: n, name, line: position.line });
      });
    }

    ts.forEachChild(n, walk);
  };

  walk(node);
  return scope;
}
