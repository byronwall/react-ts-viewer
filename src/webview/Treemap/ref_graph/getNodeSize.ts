import type { ScopeNode } from "../../../types";

// Helper function to get node size from its ID

export function getNodeSize(node: ScopeNode): number {
  if (!node || !node.id) return Infinity;
  const parts = node.id.split(":");
  if (parts.length > 1) {
    const range = parts[parts.length - 1];
    if (range) {
      const rangeParts = range.split("-");
      if (rangeParts.length === 2 && rangeParts[0] && rangeParts[1]) {
        const start = parseInt(rangeParts[0], 10);
        const end = parseInt(rangeParts[1], 10);
        if (!isNaN(start) && !isNaN(end)) {
          return end - start;
        }
      }
    }
  }
  return Infinity; // If no range, treat as largest
}
