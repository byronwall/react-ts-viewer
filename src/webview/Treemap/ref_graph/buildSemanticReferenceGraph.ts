import type { ScopeNode } from "../../../types";
import { analyzeBOI } from "./analyzeBOI";
import { findInnermostNodeByOffset } from "./graph_nodes";
import { findNodeById } from "./graph_nodes";
import { buildHierarchicalStructure } from "./buildHierarchicalStructure";
import { findNodesByName } from "./graph_nodes";
import { getNodeSize } from "./getNodeSize";
import { nodeDeclaresIdentifier } from "./ts_ast";

export interface SemanticReference {
  name: string;
  type:
    | "function_call"
    | "variable_reference"
    | "import"
    | "property_access"
    | "destructured_variable";
  sourceNodeId: string;
  targetNodeId?: string;
  targets?: string[];
  position: { line: number; character: number };
  /** Absolute character offset (from SourceFile) where this reference occurs */
  offset: number;
  /** ID of the innermost ScopeNode that contains the reference usage (filled later) */
  usageNodeId?: string;
  isInternal: boolean; // true if declared within BOI scope
  direction: "outgoing";
}

export function buildSemanticReferenceGraph(
  focusNode: ScopeNode,
  rootNode: ScopeNode
): {
  nodes: ScopeNode[];
  references: SemanticReference[];
  hierarchicalRoot: ScopeNode;
} {
  /*
  ============================================================
   buildSemanticReferenceGraph – High-Level Summary & Review
  ============================================================
  Purpose
  -------
  Given a `focusNode` (the slice of AST the user clicked on) and the
  `rootNode` (the full parsed file), this routine:
  1. Runs a "Breadth-Of-Interest" analysis (`analyzeBOI`) to collect
     external, incoming, and recursive identifier references relative to the
     focus node.
  2. Heuristically filters that raw reference list down to a *small* set of
     the "most interesting" references (`maxReferences` ≤ 20) and the nodes
     that declare / use them (`maxNodes` ≤ 20).
  3. Attempts to resolve every kept reference to a concrete declaration
     node via `findNodesByName`, `nodeDeclaresIdentifier`, and custom
     category sorting.
  4. Produces three artefacts for visualisation:
      • `nodes`               – all ScopeNodes we decided are relevant
      • `references`          – the resolved semantic edges (outgoing ↔ incoming)
      • `hierarchicalRoot`    – a trimmed AST suitable for a tree-view

  Where Things Get Hokey / Fragile
  --------------------------------
  • **Hard-coded magic numbers** – `maxNodes = 20`, `maxReferences = 20`.
    These caps are baked-in; any dataset that legitimately needs more will
    simply be truncated.

  • **Brittle identifier filtering** – The `prioritizedReferences` filter:
    – Maintains a *hand-written* list (`isGenericName`) of hundreds of common
      words.  Easy to miss cases and creates maintenance burden.
    – "LikelyTextToken" regex assumes English and specific casing.
    – Length & single-character heuristics silently drop legitimate code
      (e.g. loop indices `i`, `j`).

  • **Ad-hoc declaration ranking** – Declaration categories are compared
    by `String(n.category).startsWith(c)` instead of enums.  Ordering is
    guessed, not documented.

  • **Monolithic function** – ~300 lines mixing data collection, filtering,
    resolution, and tree construction.  Hard to test & reason about.

  • **TODO without follow-up** – Comment says "need to refactor this code
    out then write tests to verify it's working" – but it never happens.

  Quick Wins / Cleanup Ideas
  --------------------------
  1. **Extract Helpers**
      • `filterRelevantReferences(allRefs): SemanticReference[]`
      • `rankDeclarationCandidates(candidates): ScopeNode[]`
      • `resolveReference(ref, rootNode): ResolvedReference | undefined`
      Each isolated piece gets unit tests.

  2. **Parameterise Caps**
     Accept optional `{ maxNodes, maxReferences }` args so callers can tune
     behaviour instead of editing source.

  3. **Replace magic lists with Sets & config**
     Move `genericIdentifiers` and `htmlAttributeNames` into a JSON / util
     file.  Use `new Set([...])` for O(1) lookups.

  4. **Introduce TypeScript enums** for `SemanticReference.type` and
     declaration categories to avoid typo bugs and improve readability.

  5. **Leverage real AST checks**
     Instead of regex-based `isLikelyTextToken`, walk the parent JSX/Text
     nodes to know if a token originated from plain text.

  6. **Early exits for performance**
     Short-circuit inside the `for` loop once `referencedNodes` hits
     `maxNodes` *and* every remaining reference would only add more nodes.

  Testing Strategy
  ----------------
  • **Unit tests per helper** – Feed synthetic AST fragments to make sure
    filtering and ranking behave predictably (e.g. "i" is kept when inside
    a `for` loop but dropped in JSX text).

  • **Golden-file snapshots** – For a handful of real components, snapshot
    the `{ nodes, references }` JSON and ensure it stays stable.

  • **Property-based tests** – Randomly generate variable names to verify
    generic-identifier filter never drops legal identifiers of varying
    lengths/cases.

  Refactor Roadmap
  ----------------
  Phase 1: Extract helpers & add tests (no behaviour change).
  Phase 2: Replace hard-coded lists with injectable config.
  Phase 3: Split semantic-graph building and hierarchical-tree building
           into two discrete public functions.
  Phase 4: Re-introduce smarter text-token detection using real AST info.

  ============================================================
   End of meta commentary – original source follows ↓
  ============================================================
  */
  // Check if focus node has meaningful source code for analysis
  if (
    !focusNode.source ||
    typeof focusNode.source !== "string" ||
    focusNode.source.trim().length < 10
  ) {
    return {
      nodes: [focusNode],
      references: [],
      hierarchicalRoot: focusNode,
    };
  }

  // Perform full BOI analysis for all cases (including JSX)
  const boiAnalysis = analyzeBOI(focusNode, rootNode);

  const allReferences = boiAnalysis.externalReferences;

  // MUCH more restrictive filtering for focused analysis
  const maxNodes = 20; // Increase to accommodate individual variable declarations

  const referencedNodes: ScopeNode[] = [focusNode]; // Always include the focus node
  const resolvedReferences: SemanticReference[] = [];

  // Resolve semantic references to actual nodes (with strict limits)
  for (const ref of allReferences) {
    const matchingNodes = findNodesByName(rootNode, ref.name);

    if (matchingNodes.length > 0) {
      // Prefer real declaration nodes over incidental matches (e.g. an "if" clause that merely mentions the name)
      const declarationCategories = [
        "Variable",
        "Parameter",
        "Function",
        "Import",
        "Class",
      ];

      const declarationCandidates = matchingNodes.filter(
        (n) =>
          declarationCategories.includes(n.category) ||
          nodeDeclaresIdentifier(n, ref.name)
      );

      // ------------------------------------------------------------------
      // Prioritize candidates that ACTUALLY declare the identifier and whose
      // category suggests a real declaration site (e.g. Parameter, Variable,
      // Function).  If several such candidates exist, prefer the *smallest*
      // node slice (most specific).  Only if **none** declare the identifier
      // do we fall back to incidental matches (e.g. a JSX element that merely
      // references the variable).
      // ------------------------------------------------------------------
      const declPriorityOrder = [
        "Parameter",
        "Variable",
        "Function",
        "ArrowFunction",
        "Method",
        "Import",
        "Class",
      ];

      const categorize = (n: ScopeNode): number => {
        const idx = declPriorityOrder.findIndex((c) =>
          String(n.category).startsWith(c)
        );
        return idx === -1 ? declPriorityOrder.length : idx;
      };

      const poolToSort =
        declarationCandidates.length > 0
          ? declarationCandidates
          : matchingNodes;

      const sortedMatchingNodes = poolToSort.sort((a, b) => {
        // 1) Declaration category priority
        const catA = categorize(a);
        const catB = categorize(b);
        if (catA !== catB) return catA - catB;

        // 2) Slice size (smaller → more specific)
        return getNodeSize(a) - getNodeSize(b);
      });

      const specificDeclarationNode = sortedMatchingNodes[0];

      if (!specificDeclarationNode) {
        continue;
      }

      if (referencedNodes.length >= maxNodes) {
        break;
      }

      // For outgoing and recursive references
      if (specificDeclarationNode.id !== focusNode.id) {
        const usageNodeId = ref.offset
          ? findInnermostNodeByOffset(rootNode, ref.offset)?.id || focusNode.id
          : focusNode.id;

        resolvedReferences.push({
          ...ref,
          targetNodeId: specificDeclarationNode.id,
          targets: [focusNode.id],
          usageNodeId,
        });

        // Ensure both the declaration and usage nodes are included in the graph
        if (
          !referencedNodes.some(
            (sn: ScopeNode) => sn.id === specificDeclarationNode.id
          )
        ) {
          referencedNodes.push(specificDeclarationNode);
        }

        if (
          usageNodeId &&
          !referencedNodes.some((sn: ScopeNode) => sn.id === usageNodeId)
        ) {
          const usageNode = findInnermostNodeByOffset(rootNode, ref.offset);
          if (usageNode) {
            referencedNodes.push(usageNode);
          }
        }
      }
    }
  }

  // If no edges could be resolved, fall back to minimal visualization
  if (resolvedReferences.length === 0) {
    return {
      nodes: referencedNodes,
      references: [],
      hierarchicalRoot: focusNode, // Use focus node as root for simple case
    };
  }

  // Build hierarchical structure for all nodes
  const hierarchicalRoot = buildHierarchicalStructure(
    referencedNodes,
    rootNode
  );

  return {
    nodes: referencedNodes,
    references: resolvedReferences,
    hierarchicalRoot,
  };
} // Semantic reference types
