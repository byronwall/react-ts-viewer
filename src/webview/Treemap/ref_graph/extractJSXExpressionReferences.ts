import * as ts from "typescript";

import { VariableScope } from "./graph_nodes";
import { getLineAndCharacter } from "./ts_ast";
import { isVariableDeclaredInScope } from "./graph_nodes";
import { isIdentifierTypePosition, isKeyword } from "./ts_ast";
import { SemanticReference } from "./buildSemanticReferenceGraph";

// Helper function to extract references from JSX expressions

export function extractJSXExpressionReferences(
  expression: ts.Expression,
  sourceNodeId: string,
  boiScope: VariableScope,
  references: SemanticReference[],
  sourceFile: ts.SourceFile
) {
  function visitExpression(expr: ts.Expression) {
    const position = getLineAndCharacter(sourceFile, expr.pos);

    if (ts.isIdentifier(expr) && !isIdentifierTypePosition(expr)) {
      const name = expr.text;
      if (!isKeyword(name)) {
        const isInternal = isVariableDeclaredInScope(name, boiScope);

        if (!isInternal) {
          references.push({
            name,
            type: "variable_reference",
            sourceNodeId,
            position,
            offset: expr.pos,
            isInternal,
            direction: "outgoing",
            targets: [],
          });
        }
      }
    } else if (ts.isPropertyAccessExpression(expr)) {
      // Capture only the root object identifier of the JSX property access
      const rootObj = expr.expression;
      if (ts.isIdentifier(rootObj)) {
        const name = rootObj.text;
        const isInternal = isVariableDeclaredInScope(name, boiScope);
        if (!isInternal) {
          references.push({
            name,
            type: "variable_reference",
            sourceNodeId,
            position,
            offset: expr.pos,
            isInternal,
            direction: "outgoing",
            targets: [],
          });
        }
      }
    } else if (ts.isCallExpression(expr)) {
      if (ts.isIdentifier(expr.expression)) {
        const name = expr.expression.text;
        const isInternal = isVariableDeclaredInScope(name, boiScope);
        if (!isInternal) {
          references.push({
            name,
            type: "function_call",
            sourceNodeId,
            position,
            offset: expr.pos,
            isInternal,
            direction: "outgoing",
            targets: [],
          });
        }
      } else if (ts.isPropertyAccessExpression(expr.expression)) {
        const objectName = ts.isIdentifier(expr.expression.expression)
          ? expr.expression.expression.text
          : "unknown";
        const propertyName = expr.expression.name.text;
        const name = `${objectName}.${propertyName}`;
        const isInternal = isVariableDeclaredInScope(objectName, boiScope);
        if (!isInternal) {
          references.push({
            name,
            type: "property_access",
            sourceNodeId,
            position,
            offset: expr.pos,
            isInternal,
            direction: "outgoing",
            targets: [],
          });
        }
      }
    }

    // Recursively visit child expressions
    ts.forEachChild(expr, (child) => {
      if (ts.isExpression(child)) {
        visitExpression(child);
      }
    });
  }

  visitExpression(expression);
}
