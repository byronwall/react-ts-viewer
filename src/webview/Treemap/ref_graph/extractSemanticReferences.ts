import * as ts from "typescript";
import { type SemanticReference } from "./buildSemanticReferenceGraph";
import { type VariableScope } from "./graph_nodes";
import { getLineAndCharacter } from "./ts_ast";
import { isVariableDeclaredInScope } from "./graph_nodes";
import { extractJSXExpressionReferences } from "./extractJSXExpressionReferences";
import { getRangeFromNodeId } from "./graph_nodes";
import { isJSXComponentName } from "./ts_ast";
import {
  isIdentifierInsideJsxText,
  isIdentifierTypePosition,
  isKeyword,
  isPartOfDeclaration,
} from "./ts_ast";

// Extract semantic references from AST

export function extractSemanticReferences(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  sourceNodeId: string,
  boiScope: VariableScope
): SemanticReference[] {
  const references: SemanticReference[] = [];

  function visitNode(n: ts.Node) {
    const position = getLineAndCharacter(sourceFile, n.pos);

    // Function calls
    if (ts.isCallExpression(n) && ts.isIdentifier(n.expression)) {
      const name = n.expression.text;
      const isInternal = isVariableDeclaredInScope(name, boiScope);
      if (!isInternal) {
        references.push({
          name,
          type: "function_call",
          sourceNodeId,
          position,
          offset: n.pos,
          isInternal,
          direction: "outgoing",
          targets: [],
        });
      }
    }

    // Property access (calls or plain) â€“ capture ONLY the root object identifier
    const handlePropertyAccess = (pa: ts.PropertyAccessExpression) => {
      const root = pa.expression;
      if (ts.isIdentifier(root)) {
        const name = root.text;
        if (!isKeyword(name)) {
          const isInternal = isVariableDeclaredInScope(name, boiScope);
          if (!isInternal) {
            references.push({
              name,
              type: "variable_reference",
              sourceNodeId,
              position,
              offset: pa.pos,
              isInternal,
              direction: "outgoing",
              targets: [],
            });
          }
        }
      }
    };

    // Property access in a call expression: foo.bar()
    if (ts.isCallExpression(n) && ts.isPropertyAccessExpression(n.expression)) {
      handlePropertyAccess(n.expression);
    }

    // Stand-alone property access: foo.bar
    if (
      ts.isPropertyAccessExpression(n) &&
      !ts.isCallExpression(n.parent) /* avoid double count */
    ) {
      handlePropertyAccess(n);
    }

    // Variable references (not in declarations)
    if (
      ts.isIdentifier(n) &&
      !isPartOfDeclaration(n) &&
      !isIdentifierTypePosition(n)
    ) {
      // Skip identifiers that are *property names* in a property access expression
      // e.g. the `includes` in `keysToTrack.includes`. We only want the root object
      // (handled separately by `handlePropertyAccess`).
      if (ts.isPropertyAccessExpression(n.parent) && n.parent.name === n) {
        return; // ignore property name part
      }

      const name = n.text;

      // Skip common keywords and JSX component names
      if (
        !isKeyword(name) &&
        !isJSXComponentName(n) &&
        !isIdentifierInsideJsxText(n)
      ) {
        const isInternal = isVariableDeclaredInScope(name, boiScope);

        if (!isInternal) {
          references.push({
            name,
            type: "variable_reference",
            sourceNodeId,
            position,
            offset: n.pos,
            isInternal,
            direction: "outgoing",
            targets: [],
          });
        }
      }
    }

    // JSX Expression containers - extract variables from {variable} expressions
    if (ts.isJsxExpression(n) && n.expression) {
      extractJSXExpressionReferences(
        n.expression,
        sourceNodeId,
        boiScope,
        references,
        sourceFile
      );
    }

    // Import declarations
    if (ts.isImportDeclaration(n) && n.importClause) {
      if (n.importClause.name) {
        // Default import
        const name = n.importClause.name.text;
        references.push({
          name,
          type: "import",
          sourceNodeId,
          position,
          offset: n.pos,
          isInternal: false,
          direction: "outgoing",
          targets: [],
        });
      }

      if (n.importClause.namedBindings) {
        if (ts.isNamedImports(n.importClause.namedBindings)) {
          // Named imports
          n.importClause.namedBindings.elements.forEach((element) => {
            const name = element.name.text;
            references.push({
              name,
              type: "import",
              sourceNodeId,
              position,
              offset: n.pos,
              isInternal: false,
              direction: "outgoing",
              targets: [],
            });
          });
        }
      }
    }

    ts.forEachChild(n, visitNode);
  }

  visitNode(node);

  // Convert local offsets to absolute file offsets
  const focusStartOffset = getRangeFromNodeId(sourceNodeId)?.start ?? 0;
  references.forEach((r) => {
    r.offset = r.offset + focusStartOffset;
  });

  return references;
}
