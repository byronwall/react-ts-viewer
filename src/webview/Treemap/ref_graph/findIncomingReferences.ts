import type { ScopeNode } from "../../../types";
import { type SemanticReference } from "./buildSemanticReferenceGraph";
import { type VariableScope } from "./graph_nodes";
import { buildVariableScope } from "./buildVariableScope";
import { extractSemanticReferences } from "./extractSemanticReferences";
import { createSourceFile } from "./ts_ast";

// Find incoming references to the BOI from other nodes

export function findIncomingReferences(
  focusNode: ScopeNode,
  rootNode: ScopeNode,
  boiScope: VariableScope
): SemanticReference[] {
  const incomingRefs: SemanticReference[] = [];
  const boiVariableNames = Array.from(boiScope.declarations.keys());

  if (boiVariableNames.length === 0) {
    return incomingRefs;
  }

  function searchNode(node: ScopeNode) {
    // Skip the focus node itself
    if (node.id === focusNode.id) {
      return;
    }

    if (node.source && typeof node.source === "string") {
      try {
        const sourceFile = createSourceFile(node.source);
        const nodeScope = buildVariableScope(sourceFile, sourceFile);

        // Extract references from this node
        const nodeReferences = extractSemanticReferences(
          sourceFile,
          sourceFile,
          node.id,
          nodeScope
        );

        // Check if any references point to BOI variables
        nodeReferences.forEach((ref) => {
          if (boiVariableNames.includes(ref.name) && !ref.isInternal) {
            incomingRefs.push({
              ...ref,
              direction: "incoming",
              targets: [focusNode.id],
            });
          }
        });
      } catch (error) {
        console.warn(
          `⚠️ Error analyzing node ${node.id} for incoming references:`,
          error
        );
      }
    }

    // Recursively search children
    if (node.children) {
      node.children.forEach(searchNode);
    }
  }

  searchNode(rootNode);
  return incomingRefs;
}
