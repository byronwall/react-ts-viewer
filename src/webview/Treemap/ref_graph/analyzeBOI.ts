import type * as ts from "typescript";
import type { ScopeNode } from "../../../types";
import { type SemanticReference } from "./buildSemanticReferenceGraph";
import { buildVariableScope } from "./buildVariableScope";
import { extractSemanticReferences } from "./extractSemanticReferences";
import { findIncomingReferences } from "./findIncomingReferences";
import { getPathToNode } from "./graph_nodes";
import { createSourceFile, getLineAndCharacter } from "./ts_ast";

interface BOIAnalysis {
  scopeBoundary: { start: number; end: number };
  internalDeclarations: Map<
    string,
    { node: ts.Node; name: string; line: number }
  >;
  externalReferences: SemanticReference[];
  incomingReferences: SemanticReference[];
  recursiveReferences: SemanticReference[];
}

// Analyze Block of Interest (BOI) for semantic references
export function analyzeBOI(
  focusNode: ScopeNode,
  rootNode: ScopeNode
): BOIAnalysis {
  if (!focusNode.source || typeof focusNode.source !== "string") {
    console.warn("⚠️ No source code available for BOI analysis");
    return {
      scopeBoundary: { start: 0, end: 0 },
      internalDeclarations: new Map(),
      externalReferences: [],
      incomingReferences: [],
      recursiveReferences: [],
    };
  }

  try {
    // Create TypeScript source file
    const sourceFile = createSourceFile(focusNode.source);

    // Build variable scope for the BOI
    const boiScope = buildVariableScope(sourceFile, sourceFile);

    // Extract semantic references
    const allReferences = extractSemanticReferences(
      sourceFile,
      sourceFile,
      focusNode.id,
      boiScope
    );

    // ------------------------------------------------------------------
    // Compute absolute position (line/character) across the FULL source file
    // ------------------------------------------------------------------
    const fullSourceFile: ts.SourceFile | null =
      typeof rootNode.source === "string"
        ? createSourceFile(rootNode.source)
        : null;

    // Identify the variable name that *owns* the BOI (e.g. the variable to
    // which an arrow-function is assigned).  Any reference to that variable
    // is considered *internal* and should therefore be excluded from the
    // "external" set.
    const pathToFocus = getPathToNode(rootNode, focusNode.id);
    const owningVarNode = [...pathToFocus].reverse().find((n) => {
      return String(n.category) === "Variable";
    });

    const boiVarName = (() => {
      if (!owningVarNode?.label) return null;
      // Heuristic: first token of the label before whitespace or assignment
      return owningVarNode.label.split(/[\s=:{[(]/)[0] ?? null;
    })();

    // Helper to de-duplicate references (name+type+offset)
    const unique = new Map<string, SemanticReference>();
    allReferences.forEach((ref) => {
      const key = `${ref.name}|${ref.type}|${ref.offset}`;
      if (!unique.has(key)) {
        unique.set(key, ref);
      }
    });

    let dedupedReferences = Array.from(unique.values());

    // Recalculate line/character using the full file so the positions are absolute
    if (fullSourceFile) {
      dedupedReferences = dedupedReferences.map((r) => ({
        ...r,
        position: getLineAndCharacter(fullSourceFile, r.offset),
      }));
    }

    // Categorize references by direction, excluding refs that point to the
    // BOI's own variable name (if detected).
    let externalReferences = dedupedReferences.filter(
      (ref) => !ref.isInternal && ref.name !== boiVarName
    );
    const recursiveReferences = dedupedReferences.filter(
      (ref) => ref.isInternal
    );

    // Final de-duplication for external refs – one entry per (name,type)
    {
      const seen = new Map<string, SemanticReference>();
      externalReferences.forEach((ref) => {
        // not concerned with type here, just want to avoid duplicates
        const key = ref.name;
        if (!seen.has(key)) {
          seen.set(key, ref);
        }
      });
      externalReferences = Array.from(seen.values());
    }

    // Find incoming references by searching the root node for references to BOI variables
    const incomingReferences = findIncomingReferences(
      focusNode,
      rootNode,
      boiScope
    );

    return {
      scopeBoundary: { start: 0, end: focusNode.source.length },
      internalDeclarations: boiScope.declarations,
      externalReferences,
      incomingReferences,
      recursiveReferences,
    };
  } catch (error) {
    console.error("❌ Error in BOI analysis:", error);
    return {
      scopeBoundary: { start: 0, end: 0 },
      internalDeclarations: new Map(),
      externalReferences: [],
      incomingReferences: [],
      recursiveReferences: [],
    };
  }
}
