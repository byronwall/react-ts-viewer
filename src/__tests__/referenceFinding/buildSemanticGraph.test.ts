import { describe, expect, it } from "vitest";

import { buildSemanticReferenceGraph } from "../../webview/Treemap/ref_graph/buildSemanticReferenceGraph";
import { createRefGraphObjs } from "./createRefGraphObjs";

describe("buildSemanticReferenceGraph", () => {
  it("should build a semantic graph", () => {
    const { focusNode, rootNode } = createRefGraphObjs();
    const { references } = buildSemanticReferenceGraph(focusNode, rootNode);

    expect(references).toMatchSnapshot();
  });
});
