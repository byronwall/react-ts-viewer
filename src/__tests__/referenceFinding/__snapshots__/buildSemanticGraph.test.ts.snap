// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildSemanticReferenceGraph > should build a semantic graph 1`] = `
[
  {
    "direction": "outgoing",
    "isInternal": false,
    "name": "useCallback",
    "offset": 436,
    "position": {
      "character": 24,
      "line": 17,
    },
    "sourceNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/useKeyModifiers.tsx:421-695",
    "targetNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/useKeyModifiers.tsx:15-72",
    "targets": [
      "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/useKeyModifiers.tsx:421-695",
    ],
    "type": "function_call",
    "usageNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/useKeyModifiers.tsx:421-695",
  },
  {
    "direction": "outgoing",
    "isInternal": false,
    "name": "keysToTrack",
    "offset": 492,
    "position": {
      "character": 11,
      "line": 19,
    },
    "sourceNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/useKeyModifiers.tsx:421-695",
    "targetNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/useKeyModifiers.tsx:246-2102",
    "targets": [
      "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/useKeyModifiers.tsx:421-695",
    ],
    "type": "variable_reference",
    "usageNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/useKeyModifiers.tsx:488-666",
  },
  {
    "direction": "outgoing",
    "isInternal": false,
    "name": "setPressedModifiers",
    "offset": 541,
    "position": {
      "character": 60,
      "line": 19,
    },
    "sourceNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/useKeyModifiers.tsx:421-695",
    "targetNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/useKeyModifiers.tsx:329-410",
    "targets": [
      "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/useKeyModifiers.tsx:421-695",
    ],
    "type": "function_call",
    "usageNodeId": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/useKeyModifiers.tsx:488-666",
  },
]
`;
